File: dotnet/VM/StandardLibrary/StandardArrayLibrary.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Linq;

#nullable enable

namespace LysitheaVM
{
    public static partial class StandardArrayLibrary
    {
        #region Fields
        public static readonly IReadOnlyScope Scope = CreateScope();
        #endregion

        #region Methods
        public static Scope CreateScope()
        {
            var result = new Scope();

            var arrayFunctions = new Dictionary<string, IValue>
            {
                {"join", new BuiltinFunctionValue((vm, args) =>
                {
                    vm.PushStack(new ArrayValue(args.Value));
                }, "array.join")},

                {"length", new BuiltinFunctionValue((vm, args) =>
                {
                    var top = args.GetIndex<IArrayValue>(0);
                    vm.PushStack(top.ArrayValues.Count);
                }, "array.length")},

                {"set", new BuiltinFunctionValue((vm, args) =>
                {
                    var top = args.GetIndex<ArrayValue>(0);
                    var index = args.GetIndex<NumberValue>(1);
                    var value = args.GetIndex(2);
                    vm.PushStack(Set(top, index.IntValue, value));
                }, "array.set")},

                {"get", new BuiltinFunctionValue((vm, args) =>
                {
                    var top = args.GetIndex<IArrayValue>(0);
                    var index = args.GetIndex<NumberValue>(1);
                    if (top.TryGetIndex(index.IntValue, out var value))
                    {
                        vm.PushStack(value);
                    }
                    else
                    {
                        vm.PushStack(NullValue.Value);
                    }
                }, "array.get")},

                {"insert", new BuiltinFunctionValue((vm, args) =>
                {
                    var top = args.GetIndex<ArrayValue>(0);
                    var index = args.GetIndex<NumberValue>(1);
                    var value = args.GetIndex(2);
                    vm.PushStack(Insert(top, index.IntValue, value));
                }, "array.insert")},

                {"insertFlatten", new BuiltinFunctionValue((vm, args) =>
                {
                    var top = args.GetIndex<ArrayValue>(0);
                    var index = args.GetIndex<NumberValue>(1);
                    var value = args.GetIndex<IArrayValue>(2);
                    vm.PushStack(InsertFlatten(top, index.IntValue, value));
                }, "array.insertFlatten")},

                {"remove", new BuiltinFunctionValue((vm, args) =>
                {
                    var top = args.GetIndex<ArrayValue>(0);
                    var value = args.GetIndex(1);
                    vm.PushStack(Remove(top, value));
                }, "array.remove")},

                {"removeAt", new BuiltinFunctionValue((vm, args) =>
                {
                    var top = args.GetIndex<ArrayValue>(0);
                    var index = args.GetIndex<NumberValue>(1);
                    vm.PushStack(RemoveAt(top, index.IntValue));
                }, "array.removeAt")},

                {"removeAll", new BuiltinFunctionValue((vm, args) =>
                {
                    var top = args.GetIndex<ArrayValue>(0);
                    var value = args.GetIndex(1);
                    vm.PushStack(RemoveAll(top, value));
                }, "array.removeAll")},

                {"contains", new BuiltinFunctionValue((vm, args) =>
                {
                    var top = args.GetIndex<IArrayValue>(0);
                    var value = args.GetIndex(1);
                    vm.PushStack(Contains(top, value));
                }, "array.contains")},

                {"indexOf", new BuiltinFunctionValue((vm, args) =>
                {
                    var top = args.GetIndex<IArrayValue>(0);
                    var value = args.GetIndex(1);
                    vm.PushStack(IndexOf(top, value));
                }, "array.indexOf")},

                {"sublist", new BuiltinFunctionValue((vm, args) =>
                {
                    var top = args.GetIndex<ArrayValue>(0);
                    var index = args.GetIndex<NumberValue>(1);
                    var length = args.GetIndex<NumberValue>(2);
                    vm.PushStack(SubList(top, index.IntValue, length.IntValue));
                }, "array.sublist")}
            };

            result.TryDefine("array", new ObjectValue(arrayFunctions));

            return result;
        }

        public static ArrayValue Set(ArrayValue self, int index, IValue input)
        {
            var newValue = self.Value.ToList();
            newValue[self.CalcIndex(index)] = input;
            return new ArrayValue(newValue);
        }

        public static ArrayValue Insert(ArrayValue self, int index, IValue input)
        {
            var newValue = self.Value.ToList();
            newValue.Insert(self.CalcIndex(index), input);
            return new ArrayValue(newValue);
        }

        public static ArrayValue InsertFlatten(ArrayValue self, int index, IArrayValue input)
        {
            var newValue = self.Value.ToList();
            newValue.InsertRange(self.CalcIndex(index), input.ArrayValues);
            return new ArrayValue(newValue);
        }

        public static ArrayValue RemoveAt(ArrayValue self, int index)
        {
            var newValue = self.Value.ToList();
            newValue.RemoveAt(self.CalcIndex(index));
            return new ArrayValue(newValue);
        }

        public static ArrayValue Remove(ArrayValue self, IValue input)
        {
            var newValue = self.Value.ToList();
            newValue.Remove(input);
            return new ArrayValue(newValue);
        }

        public static ArrayValue RemoveAll(ArrayValue self, IValue input)
        {
            var newValue = self.Value.ToList();
            newValue.RemoveAll(v => v.CompareTo(input) == 0);
            return new ArrayValue(newValue);
        }

        public static bool Contains(IArrayValue self, IValue input)
        {
            return self.ArrayValues.Contains(input);
        }

        public static int IndexOf(IArrayValue self, IValue input)
        {
            var list = self.ArrayValues;
            for (var i = 0; i < list.Count; i++)
            {
                if (input.CompareTo(list[i]) == 0)
                {
                    return i;
                }
            }

            return -1;
        }

        public static ArrayValue SubList(ArrayValue self, int index, int length)
        {
            index = self.CalcIndex(index);
            if (length < 0)
            {
                length = self.Value.Count - index;
            }
            else
            {
                var diff = (index + length) - self.Value.Count;
                if (diff > 0)
                {
                    length -= diff;
                }
            }

            if (index == 0 && length >= self.Value.Count)
            {
                return self;
            }

            var result = new IValue[length];
            for (var i = 0; i < length; i++)
            {
                result[i] = self.Value[i + index];
            }
            return new ArrayValue(result);
        }

        public static int GeneralCompareTo(IArrayValue left, IValue? rightInput)
        {
            if (left == rightInput)
            {
                return 0;
            }

            if (left == null || !(rightInput is IArrayValue right) || left.GetType() != rightInput.GetType())
            {
                return 1;
            }

            var compareLength = left.ArrayValues.Count.CompareTo(right.ArrayValues.Count);
            if (compareLength != 0)
            {
                return compareLength;
            }

            for (var i = 0; i < left.ArrayValues.Count; i++)
            {
                var compare = left.ArrayValues[i].CompareTo(right.ArrayValues[i]);
                if (compare != 0)
                {
                    return compare;
                }
            }

            return 0;
        }

        private static bool IsAllPrimitive(IArrayValue input)
        {
            foreach (var item in input.ArrayValues)
            {
                if (!(item is BoolValue ||
                    item is NumberValue ||
                    item is StringValue))
                {
                    return false;
                }
            }

            return true;
        }

        public static string GeneralToStringSingleLine(IArrayValue input, int indent, int depth)
        {
            var result = new StringBuilder();
            result.Append('[');
            var first = true;
            foreach (var value in input.ArrayValues)
            {
                if (!first)
                {
                    result.Append(' ');
                }
                result.Append(value.ToStringFormatted(0, 0));
                first = false;
            }
            result.Append(']');
            return result.ToString();
        }

        public static string GeneralToString(IArrayValue input, int indent, int depth)
        {
            if (input.Count() < 32 && IsAllPrimitive(input))
            {
                return GeneralToStringSingleLine(input, indent, depth + 1);
            }

            var result = new StringBuilder();

            var indent1 = "";
            var indent2 = "";

            if (indent >= 0)
            {
                indent1 = " ".Repeat(indent * depth);
                indent2 = indent1 + " ".Repeat(indent);
            }

            result.Append('[');
            var first = true;
            foreach (var value in input.ArrayValues)
            {
                if (indent >= 0)
                {
                    result.Append('\n');
                    result.Append(indent2);
                }
                else if (!first)
                {
                    result.Append(' ');
                }
                first = false;

                result.Append(value.ToStringFormatted(indent, depth + 1));
            }

            if (!first && indent >= 0)
            {
                result.Append('\n');
                result.Append(indent1);
            }
            result.Append(']');
            return result.ToString();

        }
        #endregion
    }
}

File: ts/src/standardLibrary/standardArrayLibrary.ts
import Scope, { IReadOnlyScope } from "../scope";
import ArrayValue, { isArrayValue } from "../values/arrayValue";
import BuiltinFunctionValue from "../values/builtinFunctionValue";
import { IArrayValue, isIArrayValue, IValue } from "../values/ivalues";
import NullValue from "../values/nullValue";
import ObjectValue, { ObjectValueMap } from "../values/objectValue";

export const arrayScope: IReadOnlyScope = createArrayScope();

export function createArrayScope()
{
    const result = new Scope();

    const arrayFunctions: ObjectValueMap =
    {
        join: new BuiltinFunctionValue((vm, args) =>
        {
            vm.pushStack(new ArrayValue(args.value));
        }, "array.join"),

        length: new BuiltinFunctionValue((vm, args) =>
        {
            const top = args.getIndexCast(0, isArrayValue);
            vm.pushStackNumber(top.arrayValues().length);
        }, "array.length"),

        get: new BuiltinFunctionValue((vm, args) =>
        {
            const top = args.getIndexCast(0, isArrayValue);
            const index = args.getNumber(1);
            const result = get(top, index);
            if (result !== undefined)
            {
                vm.pushStack(result);
            }
            else
            {
                vm.pushStack(NullValue.Value);
            }
        }, "array.get"),

        set: new BuiltinFunctionValue((vm, args) =>
        {
            const top = args.getIndexCast(0, isArrayValue);
            const index = args.getNumber(1);
            const value = args.getIndex(2);
            vm.pushStack(set(top, index, value));

        }, "array.set"),

        insert: new BuiltinFunctionValue((vm, args) =>
        {
            const top = args.getIndexCast(0, isArrayValue);
            const index = args.getNumber(1);
            const value = args.getIndex(2);
            vm.pushStack(insert(top, index, value));
        }, "array.insert"),

        insertFlatten: new BuiltinFunctionValue((vm, args) =>
        {
            const top = args.getIndexCast(0, isArrayValue);
            const index = args.getNumber(1);
            const value = args.getIndexCast(2, isIArrayValue);
            vm.pushStack(insertFlatten(top, index, value));
        }, "array.insertFlatten"),

        remove: new BuiltinFunctionValue((vm, args) =>
        {
            const top = args.getIndexCast(0, isArrayValue);
            const value = args.getIndex(1);
            vm.pushStack(remove(top, value));
        }, "array.remove"),

        removeAt: new BuiltinFunctionValue((vm, args) =>
        {
            const top = args.getIndexCast(0, isArrayValue);
            const index = args.getNumber(1);
            vm.pushStack(removeAt(top, index));
        }, "array.removeAt"),

        removeAll: new BuiltinFunctionValue((vm, args) =>
        {
            const top = args.getIndexCast(0, isArrayValue);
            const value = args.getIndex(1);
            vm.pushStack(removeAll(top, value));
        }, "array.removeAll"),

        contains: new BuiltinFunctionValue((vm, args) =>
        {
            const top = args.getIndexCast(0, isArrayValue);
            const value = args.getIndex(1);
            vm.pushStackBool(contains(top, value));
        }, "array.contains"),

        indexOf: new BuiltinFunctionValue((vm, args) =>
        {
            const top = args.getIndexCast(0, isArrayValue);
            const value = args.getIndex(1);
            vm.pushStackNumber(indexOf(top, value));
        }, "array.indexOf"),

        sublist: new BuiltinFunctionValue((vm, args) =>
        {
            const top = args.getIndexCast(0, isArrayValue);
            const index = args.getNumber(1);
            const length = args.getNumber(2);
            vm.pushStack(sublist(top, index, length));
        }, "array.sublist"),
    }

    result.define('array', new ObjectValue(arrayFunctions));

    return result;
}

export function set(target: ArrayValue, index: number, input: IValue): ArrayValue
{
    let result = [...target.value];
    result[target.calcIndex(index)] = input;
    return new ArrayValue(result);
}

export function get(target: IArrayValue, index: number): IValue | undefined
{
    return target.tryGetIndex(index);
}

export function insert(target: ArrayValue, index: number, input: IValue): ArrayValue
{
    index = target.calcIndex(index);
    const result = [...target.value];
    result.splice(index, 0, input);
    return new ArrayValue(result);
}

export function insertFlatten(target: ArrayValue, index: number, input: IArrayValue): ArrayValue
{
    index = target.calcIndex(index);
    const result = [...target.value];
    result.splice(index, 0, ...input.arrayValues());
    return new ArrayValue(result);
}

export function removeAt(target: ArrayValue, index: number): ArrayValue
{
    index = target.calcIndex(index);
    const result = [...target.value];
    result.splice(index, 1);
    return new ArrayValue(result);
}

export function remove(target: ArrayValue, value: IValue): ArrayValue
{
    const index = indexOf(target, value);
    if (index < 0)
    {
        return target;
    }

    return removeAt(target, index);
}

export function removeAll(target: ArrayValue, value: IValue): ArrayValue
{
    return new ArrayValue(target.value.filter(v => v.compareTo(value) !== 0));
}

export function contains(target: IArrayValue, value: IValue): boolean
{
    return indexOf(target, value) >= 0;
}

export function indexOf(target: IArrayValue, value: IValue): number
{
    return target.arrayValues().findIndex(v => v.compareTo(value) === 0);
}

export function sublist(target: ArrayValue, index: number, length: number): ArrayValue
{
    index = target.calcIndex(index);
    if (length < 0)
    {
        return new ArrayValue(target.value.slice(index));
    }
    return new ArrayValue(target.value.slice(index, index + length));
}